### PR TITLE
[SPARK-50480][SQL] Extend CharType and VarcharType from StringType

### PR DESCRIPTION
**Pull request by jovanm-db**
### What changes were proposed in this pull request?
Change type hierarchy so that CharType and VarcharType extend StringType. 

### Why are the changes needed?
This change is a step towards to properly supporting char and varchar types. 

### Does this PR introduce _any_ user-facing change?
No. 

### How was this patch tested?
Added tests in `DataTypeSuite`. 

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by jovanm-db** [comment]
@mihailom-db

File: sql/api/src/main/scala/org/apache/spark/sql/types/CharType.scala [hunk 1]
```diff
@@ -17,14 +17,24 @@
 
 package org.apache.spark.sql.types
 
+import org.json4s.JsonAST.{JString, JValue}
+
 import org.apache.spark.annotation.Experimental
 
 @Experimental
-case class CharType(length: Int) extends AtomicType {
+case class CharType(length: Int) extends StringType(0, Some(length)) {
```
Comment (mihailom-db) [inline sql/api/src/main/scala/org/apache/spark/sql/types/CharType.scala:25]: Let's not put 0 here, but UTF8_BINARY constatnt

File: sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 2]
```diff
@@ -458,6 +458,10 @@ object DataType {
   private[sql] def equalsIgnoreCompatibleCollation(from: DataType, to: DataType): Boolean = {
     (from, to) match {
       // String types with possibly different collations are compatible.
+      case (CharType(l1), CharType(l2)) => l1 == l2
+      case (VarcharType(l1), VarcharType(l2)) => l1 == l2
+      case (CharType(_) | VarcharType(_), _: StringType) => false
+      case (_: StringType, CharType(_) | VarcharType(_)) => false
       case (_: StringType, _: StringType) => true
```
Comment (mihailom-db) [inline sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala:465]: @stefankandic does this seem ok to you? I do not want to break delta. Is this only used in ddl operations?

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 3]
```diff
@@ -101,7 +107,7 @@ class StringType private (val collationId: Int) extends AtomicType with Serializ
  * @since 1.3.0
  */
 @Stable
-case object StringType extends StringType(0) {
+case object StringType extends StringType(0, None) {
```
Comment (mihailom-db) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:110]: could you update 0 here as well

File: sql/api/src/main/scala/org/apache/spark/sql/types/VarcharType.scala [hunk 4]
```diff
@@ -16,14 +16,24 @@
  */
 package org.apache.spark.sql.types
 
+import org.json4s.JsonAST.{JString, JValue}
+
 import org.apache.spark.annotation.Experimental
 
 @Experimental
-case class VarcharType(length: Int) extends AtomicType {
+case class VarcharType(length: Int) extends StringType(0, Some(length)) {
```
Comment (mihailom-db) [inline sql/api/src/main/scala/org/apache/spark/sql/types/VarcharType.scala:24]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 5]
```diff
@@ -196,6 +196,8 @@ object Literal {
     case TimestampNTZType => create(0L, TimestampNTZType)
     case it: DayTimeIntervalType => create(0L, it)
     case it: YearMonthIntervalType => create(0, it)
+    case CharType(_) | VarcharType(_) =>
+      throw QueryExecutionErrors.noDefaultForDataTypeError(dataType)
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:200]: Why not add default for these types?

**Review by mihailomilosevic2001 — COMMENTED** [review]
Few small comments left.

File: sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 6]
```diff
@@ -458,6 +458,10 @@ object DataType {
   private[sql] def equalsIgnoreCompatibleCollation(from: DataType, to: DataType): Boolean = {
     (from, to) match {
       // String types with possibly different collations are compatible.
+      case (CharType(l1), CharType(l2)) => l1 == l2
+      case (VarcharType(l1), VarcharType(l2)) => l1 == l2
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala:462]: how about `(_: CharType, _: VarcharType)`?

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 7]
```diff
@@ -30,7 +30,9 @@ import org.apache.spark.sql.catalyst.util.CollationFactory
  *   The id of collation for this StringType.
  */
 @Stable
-class StringType private (val collationId: Int) extends AtomicType with Serializable {
+class StringType private[spark] (val collationId: Int, val maxLength: Option[Int] = None)
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:33]: why do we need to have the length info in the base `StringType`?

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 8]
```diff
@@ -82,8 +84,12 @@ class StringType private (val collationId: Int) extends AtomicType with Serializ
   // the collation information is written to struct field metadata
   override def jsonValue: JValue = JString("string")
 
-  override def equals(obj: Any): Boolean =
-    obj.isInstanceOf[StringType] && obj.asInstanceOf[StringType].collationId == collationId
+  override def equals(obj: Any): Boolean = {
+    obj match {
+      case s: StringType => s.collationId == collationId && s.maxLength == maxLength
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:89]: We should exclude char/varchar here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/SchemaUtils.scala [hunk 9]
```diff
@@ -317,6 +317,8 @@ private[spark] object SchemaUtils {
       StructType(fields.map { field =>
         field.copy(dataType = replaceCollatedStringWithString(field.dataType))
       })
+    case CharType(l) => CharType(l)
+    case VarcharType(l) => VarcharType(l)
     case _: StringType => StringType
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/SchemaUtils.scala:322]: can't we make this only the change in this line where we would say if it has no constraint than exchange

File: sql/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala [hunk 10]
```diff
@@ -197,6 +187,16 @@ object DataTypeProtoConverter {
...
+      case s: StringType =>
+        proto.DataType
+          .newBuilder()
+          .setString(
+            proto.DataType.String
+              .newBuilder()
+              .setCollation(CollationFactory.fetchCollation(s.collationId).collationName)
+              .build())
+          .build()
+
```
Comment (mihailom-db) [inline sql/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala:199]: Can you add comment above why we are moving this. Just something that string has to come after char/varchar

File: sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 11]
```diff
@@ -458,7 +458,7 @@ object DataType {
   private[sql] def equalsIgnoreCompatibleCollation(from: DataType, to: DataType): Boolean = {
     (from, to) match {
       // String types with possibly different collations are compatible.
-      case (_: StringType, _: StringType) => true
+      case (a: StringType, b: StringType) => a.constraint == b.constraint
```
Comment (mihailom-db) [inline sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala:461]: Don't know if this made some tests fail, but could you add tests for alter column with char/varchar to check what was behaviour before and after this change, so we do not introduce behavioural change?

**Review by mihailomilosevic2001 — COMMENTED** [review]
Please remove draft from the PR and address comments.

File: sql/api/src/main/scala/org/apache/spark/sql/types/CharType.scala [hunk 12]
```diff
@@ -17,14 +17,19 @@
...
 
 @Experimental
-case class CharType(length: Int) extends AtomicType {
+case class CharType(length: Int)
+    extends StringType(CollationFactory.UTF8_BINARY_COLLATION_ID, FixedLength(length)) {
   require(length >= 0, "The length of char type cannot be negative.")
 
   override def defaultSize: Int = length
   override def typeName: String = s"char($length)"
+  override def jsonValue: JValue = JString(typeName)
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/CharType.scala:32]: shouldn't the json value include the length?

File: sql/api/src/main/scala/org/apache/spark/sql/types/VarcharType.scala [hunk 13]
```diff
@@ -16,14 +16,19 @@
...
 
 @Experimental
-case class VarcharType(length: Int) extends AtomicType {
+case class VarcharType(length: Int)
+    extends StringType(CollationFactory.UTF8_BINARY_COLLATION_ID, MaxLength(length)) {
   require(length >= 0, "The length of varchar type cannot be negative.")
 
   override def defaultSize: Int = length
   override def typeName: String = s"varchar($length)"
+  override def jsonValue: JValue = JString(typeName)
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/VarcharType.scala:31]: ditto

**Comment by mihailom-db** [comment]
@jovanm-db update title with the component [SQL]

File: sql/api/src/main/scala/org/apache/spark/sql/types/CharType.scala (hunk 12 above)
Comment (jovanm-db) [reply sql/api/src/main/scala/org/apache/spark/sql/types/CharType.scala:32]: `typeName` includes the length

**Comment by mihailom-db** [comment]
Please merge in new master, so we do not have conflicts

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 14]
```diff
@@ -1615,9 +1615,7 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
             case (CharType(l1), CharType(l2)) => l1 == l2
             case (CharType(l1), VarcharType(l2)) => l1 <= l2
             case (VarcharType(l1), VarcharType(l2)) => l1 <= l2
-            case _ =>
-              Cast.canUpCast(from, to) ||
-                DataType.equalsIgnoreCompatibleCollation(field.dataType, newDataType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:1620]: why is this change?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:1620]: seems a reasonable cleanup but unrelated to this PR, right?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 14 above)
Comment (jovanm-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:1620]: DataType.equalsIgnoreCompatibleCollation(field.dataType, newDataType) is redundant, if it returns true then so does Cast.canUpCast(from, to)

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**